extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/base/JDBCDirectory.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.guacamole.auth.jdbc.base;
+
+import org.apache.guacamole.GuacamoleException;
+import org.apache.guacamole.net.auth.AtomicDirectoryOperation;
+import org.apache.guacamole.net.auth.Directory;
+import org.apache.guacamole.net.auth.Identifiable;
+import org.mybatis.guice.transactional.Transactional;
+
+/**
+ * An implementation of Directory that uses database transactions to guarantee
+ * atomicity for any operations supplied to tryAtomically().
+ */
```

Comment:
Does this implementation, coupled with the fact that it is used elsewhere, mean that all JDBC operations will be attempted atomically? And does that have any potential adverse impacts from a performance perspective, particularly when you have multiple people accessing the Guacamole UI, starting and stopping connections, which will involve DB reads to retrieve connection data and DB writes for history entries, etc.

Sorry if that's a stupid or obvious question...

Reply:
The only place this is used is in the [Directory PATCH endpoint](https://github.com/jmuehlner/guacamole-client/blob/GUACAMOLE-926-batch-connection-import/guacamole/src/main/java/org/apache/guacamole/rest/directory/DirectoryResource.java#L449), so it should have no effects on performance or anything else for any other JDBC call. 

So no, not all JDBC operations will be attempted atomically.